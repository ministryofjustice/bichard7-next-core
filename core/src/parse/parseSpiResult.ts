import { XMLParser } from "fast-xml-parser"
import { decodeAttributeEntitiesProcessor, decodeTagEntitiesProcessor } from "src/lib/encoding"
import type { IncomingMessageParsedXml } from "../types/SpiResult"
import { incomingMessageParsedXmlSchema } from "../types/SpiResult"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeAttributeEntitiesProcessor,
    tagValueProcessor: decodeTagEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
