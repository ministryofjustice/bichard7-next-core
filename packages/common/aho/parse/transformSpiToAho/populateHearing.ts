import type { Hearing } from "../../../types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml, SpiDefendant } from "../../../types/SpiResult"

import getOrganisationUnit from "../../getOrganisationUnit"
import removeSeconds from "./removeSeconds"

type DefendantDetailsData = {
  name: string
  presentAtHearing: string
}

const getDefendantDetails = (spiDefendant: SpiDefendant): DefendantDetailsData => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonFamilyName: familyName, PersonGivenName1: givenName }
          }
        },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant

    const name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
    return { name, presentAtHearing }
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant
    const name = spiOrganisationName.trim()
    return { name, presentAtHearing }
  }

  throw new Error("Defendant details contained neither CourtIndividualDefendant or CourtCorporateDefendant")
}

const populateHearing = (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      Case: { Defendant: spiDefendant },
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      }
    }
  } = courtResult

  const { name, presentAtHearing } = getDefendantDetails(spiDefendant)

  const hearingOutcomeHearing: Hearing = {
    CourtHearingLocation: getOrganisationUnit(spiCourtHearingLocation),
    CourtHouseCode: Number(spiPsaCode),
    DateOfHearing: new Date(spiDateOfHearing),
    DefendantPresentAtHearing: presentAtHearing,
    HearingDocumentationLanguage: "D",
    HearingLanguage: "D",
    SourceReference: {
      DocumentName: `SPI ${name}`,
      DocumentType: "SPI Case Result",
      UniqueID: messageId
    },
    TimeOfHearing: removeSeconds(spiTimeOfHearing)
  }

  return hearingOutcomeHearing
}

export default populateHearing
