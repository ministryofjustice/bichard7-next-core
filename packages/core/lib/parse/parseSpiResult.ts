import { XMLParser } from "fast-xml-parser"
import type { IncomingMessageParsedXml } from "../../types/SpiResult"
import { incomingMessageParsedXmlSchema } from "../../types/SpiResult"
import { decodeAttributeEntitiesProcessor, decodeTagEntitiesProcessor } from "../encoding"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeAttributeEntitiesProcessor,
    tagValueProcessor: decodeTagEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
