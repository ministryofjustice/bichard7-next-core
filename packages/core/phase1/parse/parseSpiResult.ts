import type { IncomingMessageParsedXml } from "core/phase1/types/SpiResult"
import { incomingMessageParsedXmlSchema } from "core/phase1/types/SpiResult"
import { XMLParser } from "fast-xml-parser"
import { decodeAttributeEntitiesProcessor, decodeTagEntitiesProcessor } from "../lib/encoding"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeAttributeEntitiesProcessor,
    tagValueProcessor: decodeTagEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
