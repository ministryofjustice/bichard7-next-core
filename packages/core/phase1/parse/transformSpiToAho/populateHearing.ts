import type { Hearing } from "../../../types/AnnotatedHearingOutcome"
import getOrganisationUnit from "../../lib/organisationUnit/getOrganisationUnit"
import type { ResultedCaseMessageParsedXml, SpiDefendant } from "../../types/SpiResult"
import removeSeconds from "./removeSeconds"

type DefendantDetailsData = {
  presentAtHearing: string
  name: string
}

const getDefendantDetails = (spiDefendant: SpiDefendant): DefendantDetailsData => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: presentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    const name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
    return { presentAtHearing, name }
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant
    const name = spiOrganisationName.trim()
    return { presentAtHearing, name }
  }
  throw new Error("Defendant details contained neither CourtIndividualDefendant or CourtCorporateDefendant")
}

const populateHearing = (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const { presentAtHearing, name } = getDefendantDetails(spiDefendant)

  const hearingOutcomeHearing: Hearing = {
    CourtHearingLocation: getOrganisationUnit(spiCourtHearingLocation),
    DateOfHearing: new Date(spiDateOfHearing),
    TimeOfHearing: removeSeconds(spiTimeOfHearing),
    HearingLanguage: "D",
    HearingDocumentationLanguage: "D",
    CourtHouseCode: Number(spiPsaCode),
    SourceReference: {
      DocumentName: `SPI ${name}`,
      DocumentType: "SPI Case Result",
      UniqueID: messageId
    },
    DefendantPresentAtHearing: presentAtHearing
  }

  return hearingOutcomeHearing
}

export default populateHearing
