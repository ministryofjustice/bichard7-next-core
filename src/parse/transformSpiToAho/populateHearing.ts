import getOrganisationUnit from "../../lib/organisationUnit/getOrganisationUnit"
import type { Hearing } from "../../types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "../../types/SpiResult"
import removeSeconds from "./removeSeconds"

export default (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const hearingOutcomeHearing = {} as Hearing

  hearingOutcomeHearing.CourtHearingLocation = getOrganisationUnit(spiCourtHearingLocation)

  hearingOutcomeHearing.DateOfHearing = new Date(spiDateOfHearing)
  hearingOutcomeHearing.TimeOfHearing = removeSeconds(spiTimeOfHearing)
  hearingOutcomeHearing.HearingLanguage = "D"
  hearingOutcomeHearing.HearingDocumentationLanguage = "D"

  let name = ""
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: spiPresentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
    name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: spiPresentAtHearing
      }
    } = spiDefendant
    name = spiOrganisationName
    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
  }
  hearingOutcomeHearing.SourceReference = {
    DocumentName: `SPI ${name}`,
    DocumentType: "SPI Case Result",
    UniqueID: messageId
  }

  hearingOutcomeHearing.CourtHouseCode = Number(spiPsaCode)

  return hearingOutcomeHearing
}
