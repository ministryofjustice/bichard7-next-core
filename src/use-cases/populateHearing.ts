import type { Hearing } from "src/types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import removeSeconds from "src/utils/removeSeconds"
import getOrganisationUnit from "./getOrganisationUnit"

export default (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const hearingOutcomeHearing = {} as Hearing

  hearingOutcomeHearing.CourtHearingLocation = getOrganisationUnit(spiCourtHearingLocation)

  hearingOutcomeHearing.DateOfHearing = spiDateOfHearing
  hearingOutcomeHearing.TimeOfHearing = removeSeconds(spiTimeOfHearing)
  hearingOutcomeHearing.HearingLanguage = "D"
  hearingOutcomeHearing.HearingDocumentationLanguage = "D"

  let name = ""
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: spiPresentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
    name = [givenName, familyName].filter((namePart) => namePart).join(" ")
  } else if (spiDefendant.CourtCorporateDefendant) {
    name = spiDefendant.CourtCorporateDefendant.OrganisationName.OrganisationName
  }
  hearingOutcomeHearing.SourceReference = {
    DocumentName: `SPI ${name}`,
    DocumentType: "SPI Case Result",
    UniqueID: messageId
  }

  hearingOutcomeHearing.CourtHouseCode = spiPsaCode

  return hearingOutcomeHearing
}
