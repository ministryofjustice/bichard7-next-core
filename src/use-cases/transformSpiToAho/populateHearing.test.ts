import { readFileSync } from "fs"
import parseSpiResult from "src/use-cases/parseSpiResult"
import populateHearing from "./populateHearing"

const message = readFileSync("test-data/input-message-001.xml", "utf-8")
const courtResult = parseSpiResult(message).DeliverRequest.Message.ResultedCaseMessage

describe("populateHearing", () => {
  it("should transform SPI Hearing to Hearing Outcome Hearing", () => {
    const messageId = "DUMMY_MESSAGE_ID"
    const hearingOutcomeCase = populateHearing(messageId, courtResult)

    expect(hearingOutcomeCase).toBeDefined()
    expect(hearingOutcomeCase).toMatchSnapshot()
  })
})
