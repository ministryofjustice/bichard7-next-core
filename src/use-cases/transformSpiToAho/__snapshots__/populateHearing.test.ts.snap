// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateHearing should transform SPI Hearing to Hearing Outcome Hearing 1`] = `
Object {
  "CourtHearingLocation": Object {
    "BottomLevelCode": "01",
    "OrganisationUnitCode": "B01EF01",
    "SecondLevelCode": "01",
    "ThirdLevelCode": "EF",
    "TopLevelCode": "B",
  },
  "CourtHouseCode": 2576,
  "DateOfHearing": 2011-09-26T00:00:00.000Z,
  "DefendantPresentAtHearing": "A",
  "HearingDocumentationLanguage": "D",
  "HearingLanguage": "D",
  "SourceReference": Object {
    "DocumentName": "SPI TRPRFOUR SEXOFFENCE",
    "DocumentType": "SPI Case Result",
    "UniqueID": "DUMMY_MESSAGE_ID",
  },
  "TimeOfHearing": "10:00",
}
`;
